### ISSUE TITLE
Realm-android not writing to ROS after renewing authentication

### ISSUE THREAD

**Issue by mhtee** [realm/realm-java#6060]
#### Goal

Realm is not writing to ROS unless I logout and re-login using username and password. I am using the fingerprint authenticator on Android. Whenever I log in through the fingerprint, I refresh and renew my auth and get a new token from Auth0. Then I use that to sign in to realm again using SyncUser.loginAsync(credentials, url, callback). The app only writes to ROS when I login using username and password. Once I exit the app and then login again using fingerprint, which uses refresh token from auth0 to get new auth token and then uses that to login to realm. The app stops writing to ROS. I also check if the syncuser is valid and it returns true. Is there something that I am doing wrong here?

```
manager.getCredentials(object : BaseCallback<Credentials, CredentialsManagerException> {
            override fun onSuccess(payload: Credentials?) {
                activity.runOnUiThread {

                    // Renew Auth Token
                    client.renewAuth(payload?.refreshToken!!)
                            .start(object : BaseCallback<Credentials, AuthenticationException> {
                                override fun onFailure(error: AuthenticationException?) {
                                    activity.runOnUiThread {
                                        onLoginFailed(error?.message!!)
                                    }
                                }

                                override fun onSuccess(credentials: Credentials?) {
                                    // Log into realm
                                    activity.runOnUiThread {
                                        val creds = SyncCredentials.jwt(credentials?.idToken!!)
                                        editor = sharedPreferences.edit()
                                        editor.putString(AUTHPREFS, credentials.idToken!!).apply()
                                        val authUrl = when {
                                            BuildConfig.DEBUG -> resources.getString(R.string.auth_url)
                                            BuildConfig.RELEASE -> resources.getString(R.string.auth_url_prod)
                                            else -> resources.getString(R.string.auth_url_test)
                                        }

                                        // Log into realm
                                        SyncUser.logInAsync(creds, authUrl, object : SyncUser.Callback<SyncUser> {
                                            /**
                                             * If the user is successfully logged into realm, dismiss the
                                             * indicator and proceed to onLoginSuccess
                                             */
                                            override fun onSuccess(result: SyncUser) {
                                                Toast.makeText(context, "LOGGED IN", Toast.LENGTH_LONG).show()
                                                var i = 0
                                                Log.d("TEST SYNCUSER", result.isValid.toString())
                                                AsyncTask.execute {
                                                    while (i < SyncUser.current().allSessions().size) {
                                                        SyncUser.current().allSessions()[i].uploadAllLocalChanges()
                                                        i++
                                                    }
                                                }
                                                onLoginSuccess()
                                            }

                                            /**
                                             * If the credentials can't log the user into realm, show the error
                                             */
                                            override fun onError(error: ObjectServerError) {
                                                val errorMsg = when (error.errorCode) {
                                                    ErrorCode.UNKNOWN_ACCOUNT -> getString(R.string.no_account_message)
                                                    ErrorCode.INVALID_CREDENTIALS -> getString(R.string.username_password_nomatch)
                                                    else -> error.toString()
                                                }
                                                onLoginFailed(errorMsg)
                                            }
                                        })
                                    }
                                }

                            })

                }
            }

            override fun onFailure(error: CredentialsManagerException?) {
                // Show error and logout
                onLoginFailed(error?.message!!)
            }

        })
```

#### Version of Realm and tooling
Realm version(s): Realm 4.3.4

Realm sync feature enabled: yes

Android Studio version: 3.1

Which Android version and device: Multiple

**Comment by nhachicha** [comment]
@mhtee once you log in the first time (using the username/password) you obtain implicitly a `refresh_token` that is used to refresh your `access_token`, if this one has expired for instance, I don't understand why you need to _refresh_ yourself the token using the `auth0` & fingerprint API after you come back to the app?

**Comment by mhtee** [comment]
@nhachicha The `refresh_token` is saved in the Auth0 CredentialsManager. I get the `refresh_token` from the credentials manager using `manager.getCredentials` function call. Once I get the refresh token then I call `client.renewAuth` to get a new idToken to log in to realm. Am I supposed to do something else.

**Comment by nhachicha** [comment]
@mhtee you don't need to do all this.

You only need to authenticate using JWT credentials, then once you get a `SyncUser` use it to build a `SyncConfiguration` to be used when you need to work with Realm. Upon the next startup of the app, you need to check if there's still a valid user, then use it otherwise show the login Activity. 

Here's an example taken from the [SyncIntro](https://github.com/realm/my-first-realm-app/tree/master/android/SyncIntro/app) app
```Java
// WelcomeActivity
@Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (SyncUser.current() != null) {
            navigateToYourMainActitiy();
        } else {
             // inflate the view to allow the user to login (password or fingerprint) 
              attemptLogin();
         }
    }

    private void attemptLogin() {
          // obtain your token from auth0
          val creds = SyncCredentials.jwt(credentials?.idToken!!)
          // Log into realm
         SyncUser.logInAsync(creds, authUrl, object : SyncUser.Callback<SyncUser> {
           override fun onSuccess(user: SyncUser) {
            // build the SyncConfiguration  
            SyncConfiguration configuration = user
                .createConfiguration(REALM_URL)
                .build();
             Realm.setDefaultConfiguration(configuration);
             navigateToYourMainActitiy();
        }
    }
 }

```
Now you can use your Realm seamlessly within the app by getting the instance via `Realm.getDefaultInstance()`

**Comment by mhtee** [comment]
I obtain the JWT credentials using `refresh_token` if the credentials are expired, because I don't store the user's username and password. And I get my SyncConfigurations lazily from a config file, where it builds the sync configuration if its null. Anyways, it is working now. I prompt the user to login (fingerprint/phone pin) every time they open the app, even if their credentials are valid because of privacy reasons. All I had to do was once the fingerprint login was successful, I had to check if my saved credentials are also valid in addition to sync user being valid.
